### ISSUE TITLE
use of XamlCompilation attribute leads to 'unable to resolve mscorlib' compliation warning

### ISSUE THREAD

**Issue by FrankChen021** [xamarin/Xamarin.Forms#1958]
### Description
if we add `[XamlCompilation(XamlCompilationOptions.Compile)]` to a Xamarin form to enable form compilation, a compilation warning shows that mscorlib conflicts as below:

> unable to resolve “mscorlib, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089” and “mscorlib, Version=2.0.5.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e” ...

### Steps to Reproduce

it's very easy to reproduce this bug
1.  create a cross-platform app(Xamarin.Forms) project with the vs2017 wizard. choose `Android` platform, and use the `.NET Standard` share strategy.
2.  by default, the wizard creates a page named as `MainPage`, open `MainPage.xaml.cs`, add `[XamlCompilation(XamlCompilationOptions.Compile)]` to the `MainPage` class
3.  compile the android project, we get the warning. and the warning will disappear if we remove that attribute

### Expected Behavior

no compilation warnings when XamlCompilation attribute is used for an android project.

### Basic Information

- Version with issue:

    currently, I use the latest version(2.5.0.280555), but this problem has been there at least since 2.4.0.280(I didn't check other versions before 2.4.0.280)

- IDE:  VS2017 on Windows 10
- Platform Target Frameworks: Android
- Android: 
- Android Support Library Version: 
    Xamarin.Android.Support.Design Version="26.1.0.1" 
    Xamarin.Android.Support.v7.AppCompat Version="26.1.0.1"
    Xamarin.Android.Support.v4 Version="26.1.0.1"
    Xamarin.Android.Support.v7.CardView Version="26.1.0.1"
    Xamarin.Android.Support.v7.MediaRouter Version="26.1.0.1"

### Reproduction Link

below is the simplest project that can reproduce this problem

[App1.zip](https://github.com/xamarin/Xamarin.Forms/files/1748732/App1.zip)

**Comment by FrankChen021** [comment]
I checked the diagnostic log of msbuild with/without `XamlCompilationOption`, the only difference is that there is a dependency to mscorlib 4.0.0 detected and it can't be resolved because currently it's built for Android, the mscorlib should be 2.0.5 shipped by Mono.

it took me some hours to dig out what happened if it's compiled with `XamlCompilationOption`.  I checked the IL DASM. i found the `InitializeComponent` of the Xamarin Form is implemented as below
```
    [GeneratedCode("Xamarin.Forms.Build.Tasks.XamlG", "0.0.0.0")]
    private void InitializeComponent()
    {
      if (ResourceLoader.get_ResourceProvider() != null && ResourceLoader.get_ResourceProvider()("App1.MainPage.xaml") != null)
        this.__InitComponentRuntime();
      else if (XamlLoader.get_XamlFileProvider() != null && XamlLoader.get_XamlFileProvider()(((object) this).GetType()) != null)
      {
        this.__InitComponentRuntime();
      }
      else
      {
        Label label = new Label();
        MainPage mainPage = this;
        NameScope nameScope = new NameScope();
        NameScope.SetNameScope((BindableObject) mainPage, (INameScope) nameScope);
        NameScope.SetNameScope((BindableObject) label, (INameScope) nameScope);
        ((BindableObject) label).SetValue((BindableProperty) Label.TextProperty, (object) "Welcome to Xamarin.Forms!");
        ((BindableObject) label).SetValue((BindableProperty) View.VerticalOptionsProperty, (object) (LayoutOptions) LayoutOptions.Center);
        ((BindableObject) label).SetValue((BindableProperty) View.HorizontalOptionsProperty, (object) (LayoutOptions) LayoutOptions.Center);
        ((BindableObject) mainPage).SetValue((BindableProperty) ContentPage.ContentProperty, (object) label);
      }
    }
```

and the IL shows that this pieces of code depends on mscorlib 4.0.0

assembly manifest:

```
// Metadata version: v4.0.30319
.assembly extern mscorlib
{
  .publickeytoken = (B7 7A 5C 56 19 34 E0 89 )                         // .z\V.4..
  .ver 4:0:0:0
}
```

```
    IL_0001: call         instance class [mscorlib]System.Func`2<string, string> [Xamarin.Forms.Core]Xamarin.Forms.Internals.ResourceLoader::get_ResourceProvider()
    IL_0006: brfalse.s    IL_0020
    IL_0008: call         instance class [mscorlib]System.Func`2<string, string> [Xamarin.Forms.Core]Xamarin.Forms.Internals.ResourceLoader::get_ResourceProvider()
    IL_000d: ldstr        "App1.MainPage.xaml"
    IL_0012: callvirt     instance !1/*string*/ class [mscorlib]System.Func`2<string, string>::Invoke(!0/*string*/)
    IL_0017: brfalse.s    IL_0020
    IL_0019: ldarg.0      // this
    IL_001a: call         instance void App1.MainPage::__InitComponentRuntime()
    IL_001f: ret          
    IL_0021: call         instance class [mscorlib]System.Func`2<class [mscorlib]System.Type, string> [Xamarin.Forms.Xaml]Xamarin.Forms.Xaml.Internals.XamlLoader::get_XamlFileProvider()
    IL_0026: brfalse.s    IL_0041
    IL_0028: call         instance class [mscorlib]System.Func`2<class [mscorlib]System.Type, string> [Xamarin.Forms.Xaml]Xamarin.Forms.Xaml.Internals.XamlLoader::get_XamlFileProvider()
    IL_002d: ldarg.0      // this
    IL_002e: call         instance class [mscorlib]System.Type [mscorlib]System.Object::GetType()
    IL_0033: callvirt     instance !1/*string*/ class [mscorlib]System.Func`2<class [mscorlib]System.Type, string>::Invoke(!0/*class [mscorlib]System.Type*/)
    IL_0038: brfalse.s    IL_0041
```

**so i'm sure this problem is caused by the IL generation for `InitializeComponent`。**

i checked the source code of `XamlCTask::TryCoreCompile` which performs the IL generation for `InitializeComponent`, and found the code:

`var func = module.ImportReference(module.ImportReference(typeof(Func<string, string>))`

the code above introduces the dependency to mscorlib

according to the brief analysis above, my conclusion is that dependency to mscorlib 4.0.0.0 is introduced **because the IL generation is performed on Windows during the msbuild process, which means XamlCTask is executed in the Windows .NET runtime, which means the code typeof(Func<string,string>) will be resolved to reference the System.Func<string, string> defined in mscorlib 4.0.0.0 instead of that defined in mscorlib 2.0.5 provided by MonoAndroid**

the IL generation code which results in dependency to mscorlib first appeared in commit f304f25df2 committed by @StephaneDelcroix . so could @StephaneDelcroix take a look at this problem ?

**Comment by FrankChen021** [comment]
i also try to fix this problem, but i have not come up with a solution. i remove some the code for IL generation so that the final `InitializeComponent ` looks like

```
    [GeneratedCode("Xamarin.Forms.Build.Tasks.XamlG", "0.0.0.0")]
    private void InitializeComponent()
    {
      Label label = new Label();
      MainPage mainPage = this;
      NameScope nameScope = new NameScope();
      NameScope.SetNameScope((BindableObject) mainPage, (INameScope) nameScope);
      NameScope.SetNameScope((BindableObject) label, (INameScope) nameScope);
      ((BindableObject) label).SetValue((BindableProperty) Label.TextProperty, (object) "Welcome to Xamarin.Forms!");
      ((BindableObject) label).SetValue((BindableProperty) View.VerticalOptionsProperty, (object) (LayoutOptions) LayoutOptions.Center);
      ((BindableObject) label).SetValue((BindableProperty) View.HorizontalOptionsProperty, (object) (LayoutOptions) LayoutOptions.Center);
      ((BindableObject) mainPage).SetValue((BindableProperty) ContentPage.ContentProperty, (object) label);
    }
```

it's very strange that **dependency to mscorlib 4.0.0.0 still exists**. IL DASM provided by Visual Studio shows that code in the method has no dependency to mscorlib but System.Runtime

  ```
// Code size       19 (0x13)
  .maxstack  8
  IL_0000:  nop
  IL_0001:  ldarg.0
  IL_0002:  ldtoken    App1.App
  IL_0007:  call       class [System.Runtime]System.Type [System.Runtime]System.Type::GetTypeFromHandle(valuetype [System.Runtime]System.RuntimeTypeHandle)
  IL_000c:  call       !!0 [Xamarin.Forms.Xaml]Xamarin.Forms.Xaml.Extensions::LoadFromXaml<class App1.App>(!!0,
                                                                                                           class [System.Runtime]System.Type)
```

while the IL decompiled by JetBrains dotPeek(2017.2.2) shows that the System.Type is resolved to that in mscorlib

```
    // [19 13 - 19 83]
    IL_0001: ldarg.0      // this
    IL_0002: ldtoken      App1.App
    IL_0007: call         class [mscorlib]System.Type [mscorlib]System.Type::GetTypeFromHandle(valuetype [mscorlib]System.RuntimeTypeHandle)
    IL_000c: call         instance !!0/*class App1.App*/ [Xamarin.Forms.Xaml]Xamarin.Forms.Xaml.Extensions::LoadFromXaml<class App1.App>(!!0/*class App1.App*/, class [mscorlib]System.Type)
    IL_0011: pop          
```

and i also add some code during the IL generation to show the assembly reference, i get this


```
1>     Module: App1.dll
1>      Resource: App1.App.xaml... Has XamlCompilationAttribute set to Skip and not Compile... skipped
1>      Resource: App1.MainPage.xaml... 
1>    -------------
1>    AssemblyReference:System.Runtime, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a
1>    AssemblyReference:Xamarin.Forms.Xaml, Version=2.0.0.0, Culture=neutral, PublicKeyToken=null
1>    AssemblyReference:Xamarin.Forms.Core, Version=2.0.0.0, Culture=neutral, PublicKeyToken=null
1>    AssemblyReference:System.Diagnostics.Tools, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a
1>
1>       Parsing Xaml... done.
1>       Replacing MainPage.InitializeComponent ()... -----------------------------done.
1>
1>    AssemblyReference:System.Runtime, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a
1>    AssemblyReference:Xamarin.Forms.Xaml, Version=2.0.0.0, Culture=neutral, PublicKeyToken=null
1>    AssemblyReference:Xamarin.Forms.Core, Version=2.0.0.0, Culture=neutral, PublicKeyToken=null
1>    AssemblyReference:System.Diagnostics.Tools, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a
1>    AssemblyReference:mscorlib, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089
1>       Optimizing IL... done
```

look, after the execution of replacing the `InitializeComponent`( i have delete the code which uses System.Func ), reference to mscorlib 4.0.0.0 is added to the assembly

**Comment by StephaneDelcroix** [comment]
this should be fixed by #1899, but I need to confirm that.

**Comment by samhouts** [comment]
Please reopen this issue if it is still a problemin 3.0.0. Thanks!
